packages/airnode-utilities/src/evm/gas-prices/gas-oracle.ts

```diff
@@ -0,0 +1,296 @@
+  switch (gasOracleConfig.gasPriceStrategy) {
+    case 'latestBlockPercentileGasPrice':
+      return await fetchLatestBlockPercentileGasPrice(provider, gasOracleConfig);
+    case 'providerRecommendedGasPrice':
+      return await fetchProviderRecommendedGasPrice(provider, gasOracleConfig);
+    case 'providerRecommendedEip1559GasPrice':
+      return await fetchProviderRecommendedEip1559GasPrice(provider, gasOracleConfig);
+    default:
+      throw new Error('Unsupported gas price oracle strategy.');
+  }
+};
+
+// Get gas price based on gas price oracle strategies
+export const getGasPrice = async (
+  provider: Provider,
+  chainOptions: config.ChainOptions
+): Promise<LogsData<GasTarget>> => {
+  const { gasPriceOracle, fulfillmentGasLimit } = chainOptions;
+  const startTime = Date.now();
+
+  const goProcessGasPriceOracleStrategies = await go(
+    () => processGasPriceOracleStrategies(provider, gasPriceOracle, startTime),
+    { totalTimeoutMs: GAS_ORACLE_STRATEGY_MAX_TIMEOUT_MS }
+  );
+
+  // Return the constant strategy gas price if all other strategies fail
+  if (!goProcessGasPriceOracleStrategies.success || !goProcessGasPriceOracleStrategies.data[1]) {
+    const constantGasPriceConfig = gasPriceOracle.find(
+      (strategy) => strategy.gasPriceStrategy === 'constantGasPrice'
+    ) as config.ConstantGasPriceStrategy;
+    const constantGasTarget = fetchConstantGasPrice(constantGasPriceConfig);
+
+    const log = logger.pend(
+      'INFO',
+      `All oracle strategies failed to return a gas price, returning constant gas price set to ${ethers.utils.formatUnits(
+        constantGasTarget.gasPrice,
+        'gwei'
+      )} gwei.`
+    );
+    const strategyLogs = goProcessGasPriceOracleStrategies.success ? goProcessGasPriceOracleStrategies.data[0] : [];
+
+    return [[log, ...strategyLogs], getGasTargetWithGasLimit(constantGasTarget, fulfillmentGasLimit)];
+  }
+
+  const [logs, strategyGasTarget] = goProcessGasPriceOracleStrategies.data;
+
+  return [logs, getGasTargetWithGasLimit(strategyGasTarget, fulfillmentGasLimit)];
+};
+
+export const processGasPriceOracleStrategies = async (
+  provider: Provider,
+  gasPriceOracleConfig: config.GasPriceOracleConfig,
+  startTime: number
+): Promise<LogsData<GasTarget | null>> => {
+  const logs: PendingLog[] = [];
+
+  const gasPriceOracleStrategies = gasPriceOracleConfig.filter(
+    (strategy) => strategy.gasPriceStrategy !== 'constantGasPrice'
+  );
+
+  // Attempt gas oracle strategies (excluding constantGasPrice) in order
+  for (const strategy of gasPriceOracleStrategies) {
+    const goAttemptGasOraclePriceStrategy = await go(() => attemptGasOracleStrategy(provider, strategy), {
+      attemptTimeoutMs: GAS_ORACLE_STRATEGY_ATTEMPT_TIMEOUT_MS,
+      totalTimeoutMs: calculateTimeout(startTime, GAS_ORACLE_STRATEGY_MAX_TIMEOUT_MS),
+      retries: 1,
+      delay: { type: 'random' as const, minDelayMs: RANDOM_BACKOFF_MIN_MS, maxDelayMs: RANDOM_BACKOFF_MAX_MS },
+    });
```

Comment:
Hmm, this works because all the strategies contain only one blockchain call. In case of a strategy that would need to make more calls, this won't work (or it would but not ideally) because you would retry the whole strategy and not just a specific call you need to retry. I think it would be better to pass the `startTime` to each strategy letting them use it how many times they need it (calculating the timeout for each potential call), similar to how we do it in Airseeker. WDYT?

Reply:
I see that makes sense, I'll make the changes :+1: